### PR TITLE
[8.0][ADD] New module quality_control

### PR DESCRIPTION
**Pull request by oihane**

### CODE REVIEW

File: quality_control/models/qc_inspection.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        string='External notes',
+        states={'success': [('readonly', True)],
+                'failed': [('readonly', True)]})
+    state = fields.Selection(
+        [('draft', 'Draft'),
+         ('ready', 'Ready'),
+         ('waiting', 'Waiting supervisor approval'),
+         ('success', 'Quality success'),
+         ('failed', 'Quality failed'),
+         ('canceled', 'Cancelled')],
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/cancelled/canceled

File: quality_control/models/qc_inspection.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        string='State', readonly=True, default='draft')
+    success = fields.Boolean(
+        compute="_success", string='Success',
+        help='This field will be marked if all tests have been succeeded.',
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: have succeeded

File: quality_control/models/qc_inspection.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        store=True)
+    auto_generated = fields.Boolean(
+        string='Auto-generated', readonly=True, copy=False,
+        help='If an inspection is auto-generated, it can be cancelled nor '
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: canceled

File: quality_control/models/qc_inspection.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+    @api.multi
+    def unlink(self):
+        for inspection in self:
+            if inspection.auto_generated:
+                raise exceptions.Warning(
+                    _("You cannot remove an auto-generated inspection"))
+            if inspection.state != 'draft':
+                raise exceptions.Warning(
+                    _("You cannot remove an inspection that it's not in draft "
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: that is not in draft

File: quality_control/models/qc_inspection.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    @api.multi
+    def action_draft(self):
+        self.write({'state': 'draft'})
+
+    @api.multi
+    def action_todo(self):
+        for inspection in self:
+            if not inspection.test:
+                raise exceptions.Warning(
+                    _("You must set the test to perform first."))
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: you must first set the test to perform.

File: quality_control/models/qc_inspection.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        comodel_name="product.product", related="inspection_id.product",
+        store=True)
+    test_line = fields.Many2one(
+        comodel_name='qc.test.question', string='Test question',
+        readonly=True)
+    possible_ql_values = fields.Many2many(
+        comodel_name='qc.test.question.value', string='Answers')
+    quantitative_value = fields.Float(
+        'Quantitative value', digits=(16, 5),
+        help="Value of the result if it's a quantitative question.")
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_inspection.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    qualitative_value = fields.Many2one(
+        comodel_name='qc.test.question.value', string='Qualitative value',
+        help="Value of the result if it's a qualitative question.",
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_inspection.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        domain="[('id', 'in', possible_ql_values[0][2])]")
+    notes = fields.Text(string='Notes')
+    min_value = fields.Float(
+        string='Min', digits=(16, 5), readonly=True,
+        help="Minimum valid value if it's a quantitative question.")
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_inspection.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    max_value = fields.Float(
+        string='Max', digits=(16, 5), readonly=True,
+        help="Maximum valid value if it's a quantitative question.")
+    test_uom_id = fields.Many2one(
+        comodel_name='product.uom', string='Test UoM', readonly=True,
+        help="UoM for minimum and maximum values if it's a quantitative "
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_inspection.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        comodel_name='qc.test.question.value', string='Qualitative value',
+        help="Value of the result if it's a qualitative question.",
+        domain="[('id', 'in', possible_ql_values[0][2])]")
+    notes = fields.Text(string='Notes')
+    min_value = fields.Float(
+        string='Min', digits=(16, 5), readonly=True,
+        help="Minimum valid value if it's a quantitative question.")
+    max_value = fields.Float(
+        string='Max', digits=(16, 5), readonly=True,
+        help="Maximum valid value if it's a quantitative question.")
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_inspection.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        comodel_name='product.uom', string='Test UoM', readonly=True,
+        help="UoM for minimum and maximum values if it's a quantitative "
+             "question.")
+    test_uom_category = fields.Many2one(
+        comodel_name="product.uom.categ", related="test_uom_id.category_id",
+        store=True)
+    uom_id = fields.Many2one(
+        comodel_name='product.uom', string='UoM',
+        domain="[('category_id', '=', test_uom_category)]",
+        help="UoM of the inspection value if it's a quantitative question.")
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: s/if it's/for

File: quality_control/models/qc_test.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from openerp import models, fields, api, exceptions, _
+
+
+class QcTest(models.Model):
+    """A test is a group of questions to with the values that make them valid.
```
Comment (elicoidal) [inline quality_control/models/qc_test.py]: s/to with/along with

File: quality_control/models/qc_test.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    @api.one
+    @api.constrains('ql_values')
+    def _check_valid_answers(self):
+        if self.type == 'quantitative':
+            return
+        for value in self.ql_values:
+            if value.ok:
+                return
+        raise exceptions.Warning(
+            _("There isn't any value with OK marked. You have to mark at "
```
Comment (elicoidal) [inline quality_control/models/qc_test.py]: s/There isn't any value with OK marked/There isn't no value marked as OK

File: quality_control/models/qc_test.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        string='Qualitative values', copy=True)
+    notes = fields.Text(string='Notes')
+    min_value = fields.Float(string='Min', digits=(16, 5))
+    max_value = fields.Float(string='Max', digits=(15, 5))
+    uom_id = fields.Many2one(comodel_name='product.uom', string='Uom')
+
+
+class QcTestQuestionValue(models.Model):
+    _name = 'qc.test.question.value'
+    _description = 'Possible values of qualitative questions.'
```
Comment (elicoidal) [inline quality_control/models/qc_test.py]: s/of/for

File: quality_control/models/qc_trigger_line.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+    trigger = fields.Many2one(comodel_name="qc.trigger", required=True)
+    test = fields.Many2one(comodel_name="qc.test", required=True)
+    user = fields.Many2one(
+        comodel_name='res.users', string='Responsible',
+        track_visibility='always', default=lambda self: self.env.user)
+    partners = fields.Many2many(
+        comodel_name='res.partner', string='Partners',
+        help='If filled, the test will only be created when the action is done'
+        ' for one of the specified partners. If empty, the test will be always'
```
Comment (elicoidal) [inline quality_control/models/qc_trigger_line.py]: s/be always/always be

File: quality_control/wizard/qc_test_wizard.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from openerp import models, fields, api
+
+
+class QcInspectionSetTest(models.TransientModel):
+    """This wizard is responsible for setting the test for a given
```
Comment (elicoidal) [inline quality_control/wizard/qc_test_wizard.py]: s/responsible for setting/is used to preset

File: quality_control/models/qc_inspection.py [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+         ('canceled', 'Cancelled')],
+        string='State', readonly=True, default='draft')
+    success = fields.Boolean(
+        compute="_success", string='Success',
+        help='This field will be marked if all tests have been succeeded.',
+        store=True)
+    auto_generated = fields.Boolean(
+        string='Auto-generated', readonly=True, copy=False,
+        help='If an inspection is auto-generated, it can be cancelled nor '
+             'removed')
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: add period

File: quality_control/models/qc_inspection.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+        if vals.get('name', '/') == '/':
+            vals['name'] = self.env['ir.sequence'].get('qc.inspection')
+        return super(QcInspection, self).create(vals)
+
+    @api.multi
+    def unlink(self):
+        for inspection in self:
+            if inspection.auto_generated:
+                raise exceptions.Warning(
+                    _("You cannot remove an auto-generated inspection"))
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: add period

File: quality_control/models/qc_inspection.py [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+            if inspection.state != 'draft':
+                raise exceptions.Warning(
+                    _("You cannot remove an inspection that it's not in draft "
+                      "state"))
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py]: add period

File: quality_control/models/qc_inspection.py [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+            if inspection.success:
+                inspection.state = 'success'
+            else:
+                inspection.state = 'waiting'
+
+    @api.multi
+    def action_approve(self):
+        for inspection in self:
+            if inspection.success:
+                inspection.state = 'success'
```
Comment (elicoidal) [inline quality_control/models/qc_inspection.py:139]: 'successful' would be better even if this is a technical status.

Comment (pedrobaeza) [inline quality_control/models/qc_inspection.py:139]: This change implies too much (possible errors replacing code+data+views, migration scripts...) for the worth.

Comment (oihane) [inline quality_control/models/qc_inspection.py:139]: @elicoidal taking into account what @pedrobaeza says I've already done all other changes

Comment (elicoidal) [inline quality_control/models/qc_inspection.py:139]: OK: being behind the scene, makes it "OK"

**Comment by agaldona** [comment]
:+1:

**Comment by pedrobaeza** [comment]
:+1:

**Comment by Daniel-CA** [comment]
:+1:

**Comment by oihane** [comment]
@pedrobaeza @elicoidal Is it now ready for merge??

**Comment by elicoidal** [comment]
+1

**Pull request merged**